opm/autodiff/StandardWell_impl.hpp

```diff
@@ -634,6 +634,71 @@ namespace Opm
                     well_state.perfPhaseRates()[(first_perf_ + perf) * np + ebosCompIdxToFlowCompIdx(componentIdx)] = cq_s[componentIdx].value();
                 }
             }
+            if (GET_PROP_VALUE(TypeTag, EnableEnergy)) {
+
+                auto fs = intQuants.fluidState();
+                int reportStepIdx = ebosSimulator.episodeIndex();
+
+                for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) {
+                    if (!FluidSystem::phaseIsActive(phaseIdx)) {
+                        continue;
+                    }
+
+                    const unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::solventComponentIndex(phaseIdx));
+                    // convert to reservoar conditions
+                    EvalWell cq_r_thermal = 0.0;
+                    if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+
+                        if(FluidSystem::waterPhaseIdx == phaseIdx)
+                             cq_r_thermal = cq_s[activeCompIdx] / extendEval(fs.invB(phaseIdx));
+
+                        // remove dissolved gas and vapporized oil
+                        const unsigned oilCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
+                        const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
+                        // q_os = q_or * b_o + rv * q_gr * b_g
+                        // q_gs = q_gr * g_g + rs * q_or * b_o
+                        // d = 1.0 - rs * rv
+                        EvalWell d = extendEval(1.0 - fs.Rv() * fs.Rs());
+                        // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
+                        if(FluidSystem::gasPhaseIdx == phaseIdx)
+                            cq_r_thermal = (cq_s[gasCompIdx] - extendEval(fs.Rs()) * cq_s[oilCompIdx]) / (d * extendEval(fs.invB(phaseIdx)) );
+                        // q_or = 1 / (b_o * d) * (q_os - rv * q_gs)
+                        if(FluidSystem::oilPhaseIdx == phaseIdx)
+                            cq_r_thermal = (cq_s[oilCompIdx] - extendEval(fs.Rv()) * cq_s[gasCompIdx]) / (d * extendEval(fs.invB(phaseIdx)) );
+
+                    } else {
+                        cq_r_thermal = cq_s[activeCompIdx] / extendEval(fs.invB(phaseIdx));
+                    }
+
+                    // change temperature for injecting fluids
+                    if (well_type_ == INJECTOR && cq_s[activeCompIdx] > 0.0){
```

Comment:
I do not know the physics equation here. Just a question here, does this mean we do not consider the production of fluids and also crossflow?

Reply:
Energy is not an "independent" quantity like the fluids, but instead it is "attached" to the something else. in the current model this "something else" is always mass (IOW we do not consider radiation) and the kinetic terms are neglected. This means that the energy source term is fully defined once you have calculated the source terms for your fluids and you know the fluid temperature. I hope this was what you were asking for.

Reply:
I suppose that if we assume to know the specific enthalpy of the injected fluid (that's a big if because that quantity depends on where the fluid came from and how it interacted with the rock while it went through the pipe), the energy part of crossflow is implicitly handled by the mass fluxes?

Reply:
> the energy part of crossflow is implicitly handled by the mass fluxes?

Nope. Crossflow is a little too complicated, let us just talk about the production. 

Will production of reservoir fluids transport some energy out?  Because ` if (well_type_ == INJECTOR && cq_s[activeCompIdx] > 0.0)` basically means we only consider the injecting of fluids through injectors.  If you think in your design, the transport of energy along production of reservoir fluids is omitted, then the current code is what you design.

Reply:
If you have some equations written somewhere, then it will also be helpful for the discussion.

Reply:
We only adjust the specific enthalpy of the fluid for injecting Injectors. For producers and producing completions in injectors we assume the fluid to have the same specific enthalpy as the perforating cell. 

```
if (well_type_ == INJECTOR && cq_s[activeCompIdx] > 0.0)
{
set the enthalpy according to the temperature of the injected fluid
}

```

So yes we consider out flow of energy for cross flowing wells. Do we do it consistently? No, since we don't keep track of the energy in the well. Is this good enough? For normal usage yes. I think so. Of course if you hit a magma channel with your injector and steam starting to flow in to your injector, then you have a big problem!

Reply:
> Of course if you hit a magma channel with your injector and steam starting to flow in to your injector, then you have a big problem!

I agree from the beginning, that crossflow is messy and it is reasonable not to handle it to some extent.  

And you are right, the production of energy is actually taken care of, I misread the the code in the bracket, sorry about that.

Reply:
> If you have some equations written somewhere, then it will also be helpful for the discussion.

basically the only thing which is relevant for the well model is the energy source term. this one is:

```
q_energy = \sum_\alpha q_\alpha*h_\alpha
```

there are some idiosyncrasies in the well model code which make the implementation a bit more complicated, but this should be the gist of the "model". do you agree @totto82?

>  Is this good enough? For normal usage yes. I think so. Of course if you hit a magma channel with your injector and steam starting to flow in to your injector, then you have a big problem!

I agree here. Also be aware that this does not only affect volcano engineering, but also normal oil reservoirs: boreholes are long and thin, so I bet a few crates of beer that the temperature of the fluid changes significantly due to heat conduction as it flows through the pipe. Lacking a proper "pipe flow model", I cannot see what can be done to capture this, though. Note that this also affects the isothermal case because resistivity of the borehole will likely be strongly dependent on fluid temperature and composition, as well as flow velocity because of turbulence within the pipe. IOW, VFP tables are a very crude way to approximate all these effects.

Reply:
> q_energy = \sum_\alpha q_\alpha*h_\alpha

Thanks for the equation. As I said, I misread the things inside the if condition, which makes some comments are noise.  Sorry about that. 

I do believe crossflow still can happen (hot water/cold water), but I always think it is messy to think about / account for. For some problems, the fluid flow back to the wellbore from the reservoir does not matter due to the fluid properties or the problem studied, for some it matters, for some, it is just reasonable not to think about it, which is a design/feature.

Since the production of energy is accounted, I have no big question about this part. I just wanted to understand the design related.

Reply:
> I do believe crossflow still can happen (hot water/cold water), exchange between layers, but I always think it is messy to think about / account for.

Yes, I agree that this is too messy at the moment. On that matter, I came to the conclusion that a proper "pipe flow" model that solves a 1-D PDE is needed do deal with this kind-of properly a while ago. This capability is way beyond what the reference simulators can do and it is also would be a major project, so it is not of immediate importance. (users seem to be happy with the crap they have. ;))